FWCore/Sources/interface/VectorInputSource.h

```diff
@@ -72,11 +72,12 @@ namespace edm {
   template<typename T>
   size_t VectorInputSource::loopOverEvents(EventPrincipal& cache, size_t& fileNameHash, size_t number, T eventOperator, CLHEP::HepRandomEngine* engine, EventID const* id, bool recycleFiles) {
     size_t i = 0U;
-    for(; i < number; ++i) {
+    while(i < number) {
       clearEventPrincipal(cache);
       bool found = readOneEvent(cache, fileNameHash, engine, id, recycleFiles);
       if(!found) break;
-      eventOperator(cache, fileNameHash);
+      bool used = eventOperator(cache, fileNameHash);
+      if(used) ++i;
```

Comment:
@makortel , here the code may stack forever  if found==true used=false, also return of i=number is not excluded.

Reply:
> return of i=number is not excluded

Probably I did not understand, but the earlier code returns `number` as well unless `readOneEvent()` returned `false`.

> here the code may stack forever if found==true used=false

That is correct, and I discussed about this in the PR description. The choice in this PR is to leave it to `eventOperator` to throw an exception if the code has entered in to a likely infinite loop (see https://github.com/cms-sw/cmssw/blob/15f86e550b119c2dbecdbd8ae643c0bc1e3753e6/SimGeneral/PreMixingModule/plugins/PreMixingModule.cc#L76-L84 for an example of such a check).

I agree it is a bit unsafe to leave such checks to `eventOperator`, but the current choice has the following benefits
* avoids tri-state bool (pileup event was used, was not used but that was OK, was not used and was not OK)
* the concrete `eventOperator` knows fully the context if a pileup event was not used, and so can include details that hopefully help in interpreting the exception message (like pointing towards a configuration error)

Reply:
@makortel , selection of an event number itself cannot take a significant CPU, may be I am wrong but to me transparency of the code in this case is more important than reduction of number of lines of code. I would not return index of the loop but result variable. May be "for" loop is indeed better than current "while". May be I am wrong, but for me "tri-state" logic is more clear even if required more logical operations. Concerning case (i==number) : it may be an old problem (hidden or overlooked) or it may never happens - not obvious from the code.

Reply:
> selection of an event number itself cannot take a significant CPU

Could you elaborate what do you mean with this?

> May be "for" loop is indeed better than current "while".

I think `while` is better because `i` is increased only selectively. With `for` I would have to decrease `i` in the non-used cases, but that works only after one pileup event has been processed (`0U-1` is a bad idea).

> I would not return index of the loop but result variable.
> ...
> Concerning case (i==number) : it may be an old problem (hidden or overlooked) or it may never happens - not obvious from the code.

The function returns the number of processed pileup events, so `i == number` is the typical case (e.g. `PileUp`, used by MixingModules, gives a warning if the number of processed pileup events differs from the requested number
https://github.com/cms-sw/cmssw/blob/c9aedc15b7d5065eb91b3708247254d23c47677e/Mixing/Base/interface/PileUp.h#L185-L187
)

Reply:
@makortel , OK , let it be as you propose.

Reply:
@civanch 
We discussed with @Dr15Jones and would propose the following
* In `VectorInputSource`, count the number of consecutive rejections of pileup events (`used == false`), and if that goes over a configurable threshold (say 100 by default), throw an exception
* In `PreMixingModule`, issue a `LogWarning` each time we get a pileup event with a true pileup value outside of the bounds of the configuration

This way the loop in `VectorInputSource` does not get stuck (for too long), and we get some detailed information (even from multiple events) to the log.

What do you think?

Reply:
Done, thanks.